src/patterns/security-code/index.md.njk

```diff
@@ -0,0 +1,135 @@
+
+Ignore any formatting or punctuation such as spaces, hyphens and dashes.
+
+Give a time limit of 15 minutes for users to enter the code. If this time limit expires, display a ‘code has expired’ error message and send the user a new code.
+
+If the user follows the ‘Not received a text message?’ link, allow them to check which mobile number they entered, and to change it if necessary. This prevents the user becoming stuck if they entered a mobile number incorrectly. Do not allow users to change the number when signing in.
+
+{{ example({group: "patterns", item: "security-code", example: "resend-first-time", html: true, nunjucks: true, open: false}) }}
+
+This should send a new code, with its own 15 minute time limit. The previous code should remain valid within its original time limit.
+
+### When signing in
+
+When a user returns to your service, verify their password first. Once they have entered this correctly, send a text message to their mobile phone with a 5 digit code in this format:
+
+{% from "govuk/components/inset-text/macro.njk" import govukInsetText %}
+{{ govukInsetText({
+  text: "12345 is your [service name] security code"
+}) }}
+
+Ask the user to enter this code. Use the same pattern and time limit as when creating an account.
+
+If they follow the ‘Not received a text message?’ link, show them a page allowing them to resend the code. Do not reveal the mobile number you sent it to.
+
+{{ example({group: "patterns", item: "security-code", example: "resend", html: true, nunjucks: true, open: false}) }}
+
+You should tell users what to do if they no longer have access to the phone used to sign up.
+
+Depending upon your service, they may need to contact a support address, or may be able to ask a colleague to update their phone number.
+
+### Domain-bound codes
+
+You can also follow the [domain-bound codes](https://developer.apple.com/news/?id=z0i801mg) standard. This improves security by associating the code with a specific domain, allowing browsers to safely autofill it. 
+
+Include the domain of the service on a new line, prefixed with an `@`, followed by a `#` symbol and the code, like this:
+
+{{ govukInsetText({
+html: "12345 is your [service name] security code<br><br>@example.com #12345"
+}) }}
+
+### Error messages
+
+Style error messages like this:
+
+{{ example({group: "patterns", item: "security-code", example: "error-incorrect", html: true, nunjucks: true, open: false}) }}
+
+If the user doesn’t enter enough digits:
+<br>Say ‘You haven’t entered enough numbers’.
+
+If the user enters too many digits:
+<br>Say ‘You’ve entered too many numbers’.
+
+If the user enters non-numeric characters, other than spaces:
+<br>Say ‘The code must be 5 numbers’.
+
+If the code has expired, show this message:
+
+{{ example({group: "patterns", item: "security-code", example: "error-expired", html: true, nunjucks: true, open: false}) }}
```

Comment:
@alexbishop1 I guess this is one of those security vs usability trade-offs that teams would have to make. To me, as a non-security expert, the risk here feels fairly small and the usability win of auto-sending a new code feels worth it?

I think this message would also not appear if the code had been successfully used – so it'd only show if the code was sent, but is being used after 15 mins have elapsed. Possibly there could also be an upper limit so that after 24 hours the code would be deleted and any would-be hacker would just see the regular "Incorrect security code" message?

There's also the suggested time delay after 10 incorrect number entries which could help mitigate?

Reply:
I agree the risk well may be small but I have not fully thought it through. It would be good to have input from a security expert. Still, it feels like it should not be impossible to confect a single incorrect code error message that also covers the possibility that the code has expired (such as, ‘Incorrect or expired security code’).

Reply:
I'm not sure how to resolve this... but Frankie's suggestion of a 'second expiration' seems sensible to address the concerns here?

We could flag this as something we need to find out more about in 'Research on this component'

Reply:
Just spoke with @christopherthomasdesign — we're going to asking around the team to make sure we're not missing a risk of some sort with this.

Reply:
@calvin-lau-sig7 thanks, please do!